### ISSUE TITLE
[Teachers Unite!] Implement optimistic whisper message

### ISSUE THREAD

**Issue by marcochavezf** [Expensify/App#26738]
After the volunteer referral flow ("I know a teacher") is completed, we redirect the user to the public room. But if the user performs this action in offline mode, we’ll just display the chat report as loading (since we’ll persist the request, we will share the room anyway once the request has been made). 

Given that’s the default offline behavior for public rooms (i.e. when the user opens the native app from a deep link), we’ll implement, as a follow-up, the optimistic whisper message to notify the user that at least something is happening.

Whisper message for the inviter/referrer:

```js
"Hey @" + inviterDisplayName + "! Thanks for referring your friend " + firstName + " to <a href=\"https://www.expensify.org\" rel=\"noreferrer noopener\">Expensify.org</a> Teachers Unite. We have gone ahead and reached out to them! We'll continue to share any important announcements here to keep you in the loop about the campaign's impact. If you'd like to refer someone else, just click on Save The World in the app's main menu. Thanks for helping us to Save The World!"
```

**Comment by rezkiy37** [comment]
I was investigating the issue. Already have an idea how to do it clear.

Guys, I’m OOO from tomorrow until next Monday due the RNEU conference. If it is not an urgent issue, I can continue to work on.

**Comment by tjferriss** [comment]
@burczu can you take a look at @hoangzinh proposal?

**Comment by burczu** [comment]
@tjferriss Yes - I'll review it today.

**Comment by burczu** [comment]
@hoangzinh Could you be more specific how exactly you want to add this `whisperedToAccountIDs` parameter to optimistic data? I've tried it with no luck...

**Comment by burczu** [comment]
Just to inform: I'll be OOO on Thursday and Friday (7-8 September).

**Comment by rezkiy37** [comment]
Hi, guys!
I am back. So, please let me know do you need my help here?

**Comment by burczu** [comment]
Not overdue - @tjferriss are we gonna assign @rezkiy37 as he is back? Or maybe we are going to put this on hold (https://github.com/Expensify/App/issues/26738#issuecomment-1712409474)?

**Comment by rezkiy37** [comment]
@tjferriss, just to friendly reminder about - https://github.com/Expensify/App/issues/26738#issuecomment-1715377224.

**Comment by tjferriss** [comment]
Yes, I think we can put this on hold for now while we figure out when we'll be launch the `Save the World` flow. The tracking issue for the project is here: https://github.com/Expensify/Expensify/issues/223645

**Comment by rezkiy37** [comment]
@tjferriss, can I continue work on the issue?
FYI: https://github.com/Expensify/App/issues/26738#issuecomment-1723750652

**Comment by rezkiy37** [comment]
Stop, @hoangzinh I don't see "Save the world" in the latest `main` anymore. Are we going to remove the flow?

<img width="650" alt="Screenshot 2023-09-19 at 14 14 33" src="https://github.com/Expensify/App/assets/57314631/ded55831-61ea-4a05-acb1-37478d54910b">
<img width="1286" alt="Screenshot 2023-09-19 at 14 14 39" src="https://github.com/Expensify/App/assets/57314631/0e0399a5-044d-4066-a8c0-1f6cae52b958">

**Comment by rezkiy37** [comment]
The issue backs to on hold - https://github.com/Expensify/App/pull/27763#issue-1902566877.

**Comment by marcochavezf** [comment]
Assigning myself to track it, I will change it to daily to check it out tomorrow

**Comment by tjferriss** [comment]
@marcochavezf have you been able to look into this one at all?

**Comment by marcochavezf** [comment]
Oops, I was focused on another wave7 item. @rezkiy37, let's continue with this issue. We're planning to add the STW button when the campaign is live, but meanwhile, let's implement it, and you can access the page by adding the route `/teachersunite` to the URL in the browser. 

@hoangzinh thanks for the proposal. When I created the issue I asked for help from the Callstack team first, appreciate your proactivity 🙇🏽

**Comment by rezkiy37** [comment]
@marcochavezf, `/teachersunite` does not work for me 🧐

<img width="1727" alt="Screenshot 2023-09-29 at 15 16 29" src="https://github.com/Expensify/App/assets/57314631/f888de9d-6f24-406a-a021-ca8f9f1d2217">

**Comment by rezkiy37** [comment]
@marcochavezf, I have a few questions:

1. Should be the whisper message interactive? I mean that in the [description](https://github.com/Expensify/App/issues/26738#issue-1881048071) there are inserts with an account and link.
2. Who should be a sender of the message? _volunteer@expensify.org_? If yes, can I hardcode login and display name of the sender?

Thank you!

**Comment by rezkiy37** [comment]
Also, can I ask you for a translation of the message?

**Comment by marcochavezf** [comment]
> @marcochavezf, `/teachersunite` does not work for me 🧐
> 
> <img alt="Screenshot 2023-09-29 at 15 16 29" width="1727" src="https://user-images.githubusercontent.com/57314631/271597046-f888de9d-6f24-406a-a021-ca8f9f1d2217.png">

Ah it should be the full relative route (without `/r`) https://staging.new.expensify.com/teachersunite

**Comment by marcochavezf** [comment]
> 1. Should be the whisper message interactive? I mean that in the https://github.com/Expensify/App/issues/26738#issue-1881048071 there are inserts with an account and link.

Yup, it should use the existing `mention` implementation which has the format `<mention-user>@${accountEmail}</mention-user>` if I recall correctly.

> 2. Who should be a sender of the message? [volunteer@expensify.org](https://www.expensify.com/_support/?supportEmail=volunteer%40expensify.org&comment=better%20supportal%20issue%20nubmer%3A%2026738&location=_support/)? If yes, can I hardcode login and display name of the sender?

Yup, you can try to "refer" a teacher with the URL posted in the previous comment and you will see the whisper that we send from the backend: 
<img width="1116" alt="Screenshot 2023-10-03 at 14 49 57" src="https://github.com/Expensify/App/assets/6829422/62aeff8e-f245-4358-abd1-b4279fd6d09a">

The idea is to create the same whisper optimistcally in the client-side instead of waiting for the server to send it

**Comment by marcochavezf** [comment]
> Also, can I ask you for a translation (ES) for the message?

Yes, and this brings an interesting point I haven't considered. Do you know if we're translating other optimistic whisper messages? Because we don't support locale in the backend atm and we're just generating this message in English, which could be a problem if the app is configured in Spanish and receives eventually the message in English. So I think we would need to modify the `ReferTeachersUniteVolunteer` API command to receive the whisper message from the client-side.

**Comment by rezkiy37** [comment]
Melvin, I am actively working on a proposal now.

**Comment by rezkiy37** [comment]
> > Also, can I ask you for a translation (ES) for the message?
> 
> Yes, and this brings an interesting point I haven't considered. Do you know if we're translating other optimistic whisper messages? Because we don't support locale in the backend atm and we're just generating this message in English, which could be a problem if the app is configured in Spanish and receives eventually the message in English. So I think we would need to modify the `ReferTeachersUniteVolunteer` API command to receive the whisper message from the client-side.

Yeah, I see that whisper messages aren't localized yet. So, I will skip localization for now. 

> So I think we would need to modify the `ReferTeachersUniteVolunteer` API command to receive the whisper message from the client-side. 

Could you please clarify what should we expect?

**Comment by rezkiy37** [comment]
A proposal is under Callstack's internal review now.

**Comment by rezkiy37** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Implement optimistic whisper message for **Teachers Unite** when a user invites someone.

### What is the root cause of that problem?
Feature request.

### What changes do you think we should make in order to solve the problem?
Basically, we need to generate a new action (_whisper message_) for the **Teachers Unite** room.

Before, we need to add a few constants for a owner of the action (_whisper message_):
1. **PUBLIC_ROOM_ADMIN_ID** - `8340048` - sender (_admin_) id.
2. **PUBLIC_ROOM_ADMIN_LOGIN** - `volunteer@expensify.org` - sender (_admin_) login.
3. **PUBLIC_ROOM_ADMIN_DISPLAY_NAME** - `Expensify.org Admin` - sender (_admin_) name.

The action includes `html` and `text` with a few inserts for a current user email and receiver email. Also, it includes `whisperedToAccountIDs` with a current user id (`to be a whisper message`).

Btw, we need to sync the front-end and backend whisper messages. Because they are the same, but with different ids.

<details><summary>Example</summary>
<p>

https://github.com/Expensify/App/assets/57314631/56705aa5-0480-4e05-929e-13b192fff3e3

</p>
</details>

**Comment by rezkiy37** [comment]
@marcochavezf, please take a look at a proposal above 🙂

**Comment by marcochavezf** [comment]
Hi @rezkiy37, yeah the proposal looks good, thanks! I think the only change we'd need to sync the frontend and backend is to send the `reportActionID` from the optimistic whisper message to the API command `ReferTeachersUniteVolunteer` as a parameter. Could you include that change so I can test the backend changes?

**Comment by rezkiy37** [comment]
> Hi @rezkiy37, yeah the proposal looks good, thanks! I think the only change we'd need to sync the frontend and backend is to send the `reportActionID` from the optimistic whisper message to the API command `ReferTeachersUniteVolunteer` as a parameter. Could you include that change so I can test the backend changes?

Added a commit - ead10a05da9d906fd81e893f45562c8a0a0354a9. I've tested and it is still duplicating a message.

**Comment by rezkiy37** [comment]
@marcochavezf, a friendly reminder about - https://github.com/Expensify/App/issues/26738#issuecomment-1755766345.

**Comment by marcochavezf** [comment]
Thanks for the change, I will working on the backend changes to avoid the duplicate message

**Comment by marcochavezf** [comment]
No update

**Comment by marcochavezf** [comment]
No update

**Comment by marcochavezf** [comment]
No update

**Comment by marcochavezf** [comment]
No update

**Comment by rezkiy37** [comment]
Waiting on the backend changes.

**Comment by rezkiy37** [comment]
No update.

**Comment by rezkiy37** [comment]
@marcochavezf, should we add "On Hold" to this issue?

**Comment by rezkiy37** [comment]
I have to inform that starting tomorrow I have a short vocation until next Monday (_6.11.2023_). Feel free to left any comments, I will address them. See you soon 😉

**Comment by burczu** [comment]
Just to inform, I'll be ooo for the next few days and we'll be back on Monday, November 6th.

**Comment by marcochavezf** [comment]
No update, seems Melvin didn't detect the on hold label in the title

**Comment by rezkiy37** [comment]
@marcochavezf, we can change the updates frequency from **daily** to **weekly**.

**Comment by marcochavezf** [comment]
Yeah, good call, I will move it to weekly

**Comment by tjferriss** [comment]
I'm asking in [Slack](https://expensify.slack.com/archives/C02NZ2HGJAZ/p1700026841448349) about the priority of this issue

**Comment by rezkiy37** [comment]
Please let us know then.

**Comment by tjferriss** [comment]
Alright, the priority is really remaining the same, but we're taking this off hold.

**Comment by rezkiy37** [comment]
Still waiting on the backend here.

**Comment by rezkiy37** [comment]
@marcochavezf, any news?

**Comment by marcochavezf** [comment]
Hi @rezkiy37, sorry for the delay here. I've been focused on other high-priority items. I plan to retake this one this week (or early next week).

**Comment by marcochavezf** [comment]
No update, but I'm retaking this one today or tomorrow.

**Comment by burczu** [comment]
Hey @tjferriss! I'll be OOO till January 6th, and right after that I'm leaving the C+ role and joining Waves. Could you please assign another C+ engineer to handle this issue from now on? Thanks.

**Comment by rezkiy37** [comment]
Guys, I am having a vacation till the 4th January. So, I will be OOO within this period.

**Comment by situchan** [comment]
I can take over as C+

**Comment by marcochavezf** [comment]
No update, we'd need to implement first a way to test the TU flows without impacting the data in the real policy. I already created a couple of PRs to flip the constants and one of those has been merged.

**Comment by JmillsExpensify** [comment]
Feel free to move this to vip-vsb if you'd like to prioritize this, though I don't think it's required for wave7 any longer.

**Comment by marcochavezf** [comment]
Moved to vip-vsb with priority Medium although it's not part of Complex splits

**Comment by tjferriss** [comment]
@marcochavezf any update here?

**Comment by marcochavezf** [comment]
Sorry no update, I've been focused on wave bugs

**Comment by marcochavezf** [comment]
> Sorry no update, I've been focused on wave bugs

Same, but fortunately most of them are done. I will move it to daily to continue with this one next week

**Comment by marcochavezf** [comment]
No update today, wrapping up other dailies, then I will continue with this one

**Comment by marcochavezf** [comment]
Phew, I thought I was going to have time to check out this one, still a bit behind with some critical bugs

**Comment by marcochavezf** [comment]
Prioritizing today some critical bugs

**Comment by strepanier03** [comment]
Looks like we need a new BZ here, reassigning.

**Comment by abekkala** [comment]
https://expensify.slack.com/archives/C01GTK53T8Q/p1708375641196609
Closing
